src/elevateprivilegeshelper.cpp

```diff
@@ -0,0 +1,186 @@
+        args.emplace_back(opt.toLocal8Bit());
+    const auto wd = workingDirectory.toLocal8Bit();
+
+    std::vector<char*> argsArray(args.size() + 1);
+    for (size_t i = 0; i < args.size(); ++i)
+        argsArray[i] = args[i].data();
+    argsArray.back() = nullptr;
+
+    // fork
+    m_pid = fork();
+
+    if (m_pid == 0) { // inside child process
+        // change working dir
+        if (!wd.isEmpty() && chdir(wd.data()) != 0)
+            qFatal("Failed to change working directory to %s", wd.data());
+
+        // stop self
+        raise(SIGSTOP);
+
+        // exec
+        execvp(argsArray[0], argsArray.data());
+        qFatal("Failed to exec %s", argsArray[0]);
+    } else if (m_pid < 0) {
+        qCritical("Failed to fork (?)");
+        return false;
+    }
+
+    return true;
+}
+
+bool InitiallyStoppedProcess::run()
+{
+    if (m_pid <= 0)
+        return false;
+
+    // wait for child to be stopped
+
+    int wstatus;
+    waitpid(m_pid, &wstatus, WUNTRACED);
+    if (!WIFSTOPPED(wstatus)) {
+        m_pid = -1;
+        return false;
+    }
+
+    // continue
+
+    ::kill(m_pid, SIGCONT);
+    return true;
+}
+
+void InitiallyStoppedProcess::terminate()
+{
+    if (m_pid > 0)
+        ::kill(m_pid, SIGTERM);
+}
+
+void InitiallyStoppedProcess::kill()
+{
+    if (m_pid > 0) {
+        ::kill(m_pid, SIGKILL);
+        waitpid(m_pid, nullptr, 0);
+        m_pid = -1;
+    }
+}
+
+PerfControlFifoWrapper::~PerfControlFifoWrapper()
+{
+    close();
+}
+
+bool PerfControlFifoWrapper::open()
+{
+    close();
+
+    QString fifoParentPath = QStandardPaths::writableLocation(QStandardPaths::RuntimeLocation);
+    if (fifoParentPath.isEmpty())
+        fifoParentPath = QStandardPaths::writableLocation(QStandardPaths::TempLocation);
+
+    const auto fifoBasePath =
+        QStringLiteral("%1/hotspot-%2-%3-perf")
+            .arg(fifoParentPath, QString::number(getpid()), QUuid::createUuid().toString().mid(1, 6));
+    m_ctlFifoPath = fifoBasePath + QStringLiteral("-control.fifo");
+    m_ackFifoPath = fifoBasePath + QStringLiteral("-ack.fifo");
+
+    if (mkfifo(m_ctlFifoPath.toLocal8Bit().data(), 0600) != 0) {
+        qCritical() << "Cannot create fifo" << m_ctlFifoPath;
+        return false;
+    }
+    if (mkfifo(m_ackFifoPath.toLocal8Bit().data(), 0600) != 0) {
+        qCritical() << "Cannot create fifo" << m_ackFifoPath;
+        return false;
+    }
+
+    m_ctlFifoFd = ::open(m_ctlFifoPath.toLocal8Bit().data(), O_RDWR);
+    if (m_ctlFifoFd < 0) {
+        qCritical() << "Cannot open fifo" << m_ctlFifoPath;
+        return false;
+    }
+    m_ackFifoFd = ::open(m_ackFifoPath.toLocal8Bit().data(), O_RDONLY | O_NONBLOCK);
+    if (m_ackFifoFd < 0) {
+        qCritical() << "Cannot open fifo" << m_ackFifoPath;
+        return false;
+    }
+
+    return true;
+}
+
+void PerfControlFifoWrapper::start()
+{
+    if (m_ctlFifoFd < 0)
+        return;
+
+    m_ackReady = new QSocketNotifier(m_ackFifoFd, QSocketNotifier::Read);
```

Comment:
is this leaked? don't you need to set a parent?

should it be asserted/verified that the m_ackReady wasn't created yet?